### ISSUE TITLE
[Core][Parameters] Adding function to check for mandatory settings

### ISSUE THREAD

**Issue by philbucher** [KratosMultiphysics/Kratos#5106]
The following came up in a discussion with @LHoeffner:
Currently `Kratos::Parametes` does not provide an interface to check if the user has specified mandatory settings in the Input, 
e.g. in a process often a `ModelPart` is being used that requires the definition of a name such that it can be extracted from the `Model` => [example](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/python_scripts/assign_scalar_variable_process.py#L39-L40)
Usually defaults are specified that throw a more or less useful error later on

I am proposing to add a new method that is called before the validation of the settings to check if some settings were specified by the user

here is what I mean in dummy-code:
~~~py
user_settings = Parameters(...) # this comes usually from json

mandatory_settings = Parameters(r"""{
    "model_part_name" : null,
    "linear_solver_settings" : {
        "solver_name" : null
}""")

# the following throws e.g. if "model_part_name" was not specified by the user 
# => basically checks if all keys in "mandatory_settings" are also in "user_settings"
user_settings.RecursivelyCheckForMandatorySettings(mandatory_settings)

# afterwards comes the validation (which also contains the typechecks)
defaults = Parameters(...)
user_settings.ValidateAndAssignDefaults(defaults) 
~~~

The proposed method should be relatively simple to implement, and the benefit is that it is clearly communicated to the user what the mandatory settings are by directly throwing meaningful errors

@KratosMultiphysics/implementation-committee @KratosMultiphysics/technical-committee

**Comment by LHoeffner** [comment]
@philbucher as you know I support the idea, however I decided for a slightly different approach in the framework for my thesis:
A: I use a specific string in the value of a mandatory setting in the format ">mandatory_[type]<" and while this might be worse style and safety to compare strings here, it does not compromise on the freedom of choosing any settings-structure. 
B: I integrated it into the 'ValidateAndAssign' function, reducing lines of code and hopefully increasing adoption. 

Even if you don't share my motivations you might want to use some dummy values (instead of null) in the mandatory settings to enable type-checking.

**Comment by philbucher** [comment]
thanks for your comments @LHoeffner , here are my answers/thoughts also for others / for the discussion
- in Kratos the types are already checked in the `AssignAndValidateDefaults` fct. Hence I don't want to introduce an additional check for the type since it is later checked again.
- I thought of option A too (passing a list of mandatory keywords, did I understand you correctly?) Unfortunately I am not sure how to get this working for a settings-structure that also has sub-settings (the "linear_solver_settings" in my initial post). Ofc in python I can have a list of multiple types, but I am not sure if it can be converted properly when passing it to C :/
I think this would be the cleanest solution, without having to specify `null` all the time, which is only needed to get a valid json
- Option B is a good suggestion! Alternatively to having the `RecursivelyCheckForMandatorySettings` separately, we could overload `AssignAndValidateDefaults` and optionally pass the mandatory settings
- Another option would be sth in the direction of #653:

    ~~~py
    user_settings = Parameters(...) # this comes usually from json

    default_settings = Parameters(r"""{
        "@mandatory_model_part_name" : "",
        "@mandatory_linear_solver_settings" : {
            "@mandatory_solver_name" : ""
        }
        "other_non_mandatory_default" : 15
    }""")

    # the following throws e.g. if "model_part_name" was not specified by the user 
    # => basically checks if all keys in that start with "@mandatory_" in the defaults 
    # are also existing in the are also in "user_settings"

    user_settings.ValidateAndAssignDefaults(defaults) 
    ~~~
    This however requires some string manipulation which is not nice

**Comment by jcotela** [comment]
Before we reinvent the wheel yet again, let me direct your attention to json schema: http://json-schema.org/understanding-json-schema/

Note that there exist extensions of nlohmann json to provide support for schema, for example https://github.com/pboettch/json-schema-validator

Also, see the (old, abandoned) discussion in #653

More to the point, if we want to keep using json for our settings, I would go towards more "standard" ways to define what a json should look like, rather than invent our own.

**Comment by philbucher** [comment]
good points @jcotela , I will check your links

BTW I read that nlohmann intentionally does NOT support more keywords/references/comments/... and other fancy/useful stuff (just for the record)

**Comment by jcotela** [comment]
@philbucher I think I can fully understand why... Also note that schema are not json proper, but a proposed extension (that's why I called them "standard" with the quotes), which is something to keep in mind before committing to it.

**Comment by LHoeffner** [comment]
I think I expressed myself in a less then ideal way @philbucher :

What I meant with A was that instead of changing anything about the structure or the keys in the settings files, mandatory settings only get a specific value (that string ">mandatory_[type]<") in the defaults. And example would look like:

```
{
    "surface_opt"     : "something",
    "some_needed_opt"  : "<mandatory_int>",
    "example_block"     : {
        "solver_type"   : "some_solver",
        "start_time"    : 0.0
    }
}
```
I also know that `AssignAndValidateDefaults` is type-checking but it requires a value in the defaults to know which type to check for (so `null` won't help here). Since my 'mandatory value' will always be a string-type in the defaults, I added another if-branch to the checking which reads the specified type from the string. You can have a look how I did it [here](https://github.com/LHoeffner/MDAOfabric/blob/38cc27640e7638f02c56a701699c03216c02c5d8/MDAOfabric/accessories/settings.py#L67) (again, I don't claim there is much beauty in that solution).

@jcotela is right that this approach is pretty far off the standard json schema (Philipp's is actually closer) but that the standard (as far as I know) is preventing type-checking and free-structuring (as explained) and I therefore decided to leave it.

**Comment by loumalouomega** [comment]
Which is the conclusion of this then?

**Comment by loumalouomega** [comment]
https://github.com/KratosMultiphysics/Kratos/pull/4905 Related

**Comment by AlejandroCornejo** [comment]
The @KratosMultiphysics/implementation-committee supports the proposal of @philbucher about implementing 2 separated methods (one to CheckDefaults and another one for enforce mandatory settings). @loumalouomega points out that ther's a danger of having to much repated code so perhaps the whole class should be revised like in #4905.

**Comment by RiccardoRossi** [comment]
my feeling is that i like more the "<mandatory_int>" format...

@jcotela what is your concrete proposal here?

**Comment by GuillermoCasas** [comment]
> I think I expressed myself in a less then ideal way @philbucher :
> 
> What I meant with A was that instead of changing anything about the structure or the keys in the settings files, mandatory settings only get a specific value (that string ">mandatory_[type]<") in the defaults. And example would look like:
> 
> ```
> {
>     "surface_opt"     : "something",
>     "some_needed_opt"  : "<mandatory_int>",
>     "example_block"     : {
>         "solver_type"   : "some_solver",
>         "start_time"    : 0.0
>     }
> }
> ```
> 
> I also know that `AssignAndValidateDefaults` is type-checking but it requires a value in the defaults to know which type to check for (so `null` won't help here). Since my 'mandatory value' will always be a string-type in the defaults, I added another if-branch to the checking which reads the specified type from the string. You can have a look how I did it [here](https://github.com/LHoeffner/MDAOfabric/blob/38cc27640e7638f02c56a701699c03216c02c5d8/MDAOfabric/accessories/settings.py#L67) (again, I don't claim there is much beauty in that solution).
> 
> @jcotela is right that this approach is pretty far off the standard json schema (Philipp's is actually closer) but that the standard (as far as I know) is preventing type-checking and free-structuring (as explained) and I therefore decided to leave it.

I see a problem here with reusing parameters. The parameters object should not know what is mandatory and what is not. Only the receiving object should, I think.

**Comment by AlejandroCornejo** [comment]
@philbucher and @LHoeffner Have you had the time to look at @jcotela suggested sources?

**Comment by pooyan-dadvand** [comment]
I like more the idea of the schema proposed by @jcotela. Nevertheless, I could not find a standard way to enforce mandatory checks by it except giving a value that is not valid. 

@jcotela, do you know what would be the correct mechanism to do so using schema?

**Comment by pooyan-dadvand** [comment]
@jcotela would you please expand more about the using the schema?

**Comment by jcotela** [comment]
@pooyan-dadvand as you say, there is no standard way of doing this. There is this schema proposal I was linking to, but I do not know if it has enough traction to be considered the definitive answer (It was the best I could find at the time, though). Essentially, it is a json-based way of describing what the json input should look like. An example from the linked json-schema-validator:
```
static json person_schema = R"(
{
    "$schema": "http://json-schema.org/draft-07/schema#",
    "title": "A person",
    "properties": {
        "name": {
            "description": "Name",
            "type": "string"
        },
        "age": {
            "description": "Age of the person",
            "type": "number",
            "minimum": 2,
            "maximum": 200
        }
    },
    "required": [
                 "name",
                 "age"
                 ],
    "type": "object"
}

)"_json;

// The people are defined with brace initialization
static json bad_person = {{"age", 42}};
static json good_person = {{"name", "Albert"}, {"age", 42}};
```

I am not saying that this is a perfect fit for us (if anything, it quickly gets verbose, as the example shows), but I suspect that having a custom solution for this would be at least similarly cumbersome.

It does have the advantage of being already implemented (the linked repository is built on top of nlohmann-json), so at least we would not need to support it from scratch. Another advantage (from my point of view) is that json schema supports references so we could, for example, define once how to validate a "matrix", a "model part name" or a "domain size" and then do it consistently across kratos.

Again, I am not pushing for this particular solution, but seeing how this kind of thing is a bit out of our regular expertise and also a problem for other users of json, I would be in support of going with any kind of standard solution before starting from scratch on our own.

**Comment by pooyan-dadvand** [comment]
Thanks @jcotela 

I think schema is indeed interesting feature and we should consider it at least to have a coherent solution with it.

**Comment by roigcarlo** [comment]
@KratosMultiphysics/technical-committee agrees on:
1.  Having a GetDefaultParameter method for the process to be added to the base class and gradually populated in the processes
2. Having a separate MetaData/Schema json (with a method to get it) describing if the field is mandatory, possible options and descriptions. This can be used in python level to inform better the developer about the use of certain complex solvers and processes like GidIO, AMGSolver, etc...
3. We try to be compatible as much as possible with schema

**Comment by loumalouomega** [comment]
>     1. Having a GetDefaultParameter method for the process to be added to the base class and gradually populated in the processes
> 

Soived in #6097

**Comment by RiccardoRossi** [comment]
@KratosMultiphysics/implementation-committee we are not clear about what is the current status right now.

Also, is there someone volunteering to implement any proposed change?

**Comment by loumalouomega** [comment]
> 
> 
> @KratosMultiphysics/implementation-committee we are not clear about what is the current status right now.
> 
> Also, is there someone volunteering to implement any proposed change?

@RiccardoRossi 

"1. Having a GetDefaultParameter method for the process to be added to the base class and gradually populated in the processes"

This is solved in #6097

**Comment by loumalouomega** [comment]
>     2. Having a separate MetaData/Schema json (with a method to get it) describing if the field is mandatory, possible options and descriptions. This can be used in python level to inform better the developer about the use of certain complex solvers and processes like GidIO, AMGSolver, etc...

Proposal for point 2 in #6590
